Star rating metafield

sofiamatulis opened the pull request:
**Why are these changes introduced?**

Fixes #0.

**What approach did you take?**

**Other considerations**

**Demo links**

- [Store](url)
- [Editor](url)

**Checklist**
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

sofiamatulis commented:
> @sofiamatulis Where can I test this?

Here is the editor link: https://os2-demo.myshopify.com/admin/themes/124585803798/editor and you can test it on the product page of `Bo Ivy - Violet` and on product cards of that product (product recommendations, featured collection)

svinkle commented:
- VO announces "4.5 out of 5.0 stars, group" then "black star black star black star"
- NVDA and Narrator skip the star group completely
- JAWS announced "black star black star black star"
- TalkBack announces "4.5 out of 5.0 stars"
- Some screen readers render an extra tab-stop due to the separation of the "20" and "total reviews" text
- The subdued icon color does not meet [WCAG 1.4.11: Non-text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) requirements of `3:1` (currently `1.35:1`)
- Due to how the star visual is generated via CSS, Windows High Contrast will always show a 5/5 rating:
  <img width="352" alt="Screen Shot 2021-08-16 at 3 46 59 PM" src="https://user-images.githubusercontent.com/1392632/129620937-65501275-bec3-4c15-bfc6-4109b2e3170a.png">

We might need to take a different approach. If it's possible to generate a series of images or use SVG, we could simply set the image `alt` text to share the review value. This would have the greatest support across environments. To better support Windows High Contrast mode, the "subdued" portion could be an hollow outline, as apposed to a filled in star ([as seen in the previous review app](https://github.com/Shopify/app_reviews/issues/1326).) Thoughts?

sofiamatulis commented:
@svinkle Thanks for the detailed review 🙏 

> * VO announces "4.5 out of 5.0 stars, group" then "black star black star black star"

I added `aria-hidden="true"` to the CSS stars. Would that make sense so it doesnt announce that?

> * NVDA and Narrator skip the star group completely

I added the `role=img` and the aria-label to the top parent, which is now wrapping all of the other elements. Does that make sense?

> * JAWS announced "black star black star black star"

Added aria-hidden true.

> * TalkBack announces "4.5 out of 5.0 stars"
> * Some screen readers render an extra tab-stop due to the separation of the "20" and "total reviews" text

I separated them and added aria-hidden true to the `p`. So it looks like this:

```
<p class="caption" aria-hidden="true">({{ product.metafields.reviews.rating_count }})</p>
 <span class="visually-hidden">{{ product.metafields.reviews.rating_count }} {{ "accessibility.total_reviews" | t }}</span>
```

> * The subdued icon color does not meet [WCAG 1.4.11: Non-text Contrast](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) requirements of `3:1` (currently `1.35:1`)

I have brought this up to Meli and Wik.

> * Due to how the star visual is generated via CSS, Windows High Contrast will always show a 5/5 rating:
>   <img alt="Screen Shot 2021-08-16 at 3 46 59 PM" width="352" src="https://user-images.githubusercontent.com/1392632/129620937-65501275-bec3-4c15-bfc6-4109b2e3170a.png">
> 

I have kept it with the CSS approach for now, but if this is causing too many accessibility issues I can change it. I have updated my PR for the product page. Let me know if this makes sense:

```
      <div class="rating" role="img" aria-label="{{ 'accessibility.star_reviews_info' | t: rating_value: product.metafields.reviews.rating.value, rating_max: product.metafields.reviews.rating.value.scale_max }}">
        <div aria-hidden="true" class="rating-star color-icon-{{ settings.accent_icons }}" style="--rating: {{ product.metafields.reviews.rating.value }}; --rating-max: {{ product.metafields.reviews.rating.value.scale_max }};"></div>
        <p class="caption" aria-hidden="true">({{ product.metafields.reviews.rating_count }})</p>
       <span class="visually-hidden">{{ product.metafields.reviews.rating_count }} {{ "accessibility.total_reviews" | t }}</span>
</div>
```
> 

> 1. If it's possible to generate a series of images or use SVG, we could simply set the image `alt` text to share the review value. This would have the greatest support across environments.
> 2. To better support Windows High Contrast mode, the "subdued" portion could be an hollow outline, as apposed to a filled in star ([as seen in the previous review app](https://github.com/Shopify/app_reviews/issues/1326).)

I have brought this up to Meli and Wik, but havent made any changes yet.

> 3. At the very least, need to add `role="img"` to the `div` to help with NVDA and Narrator to actually announce something.

✅

svinkle commented:
@sofiamatulis This is looking good. A few changes to the markup:

1. Move the reviews count out of the `role=img` container; the explicit `aria-label` overrides announcing the review count.
2. Other minor structural changes…

```html
<div class="rating" role="img" aria-label="4.5 out of 5 stars">
  <span aria-hidden="true" class="rating-star color-icon"></span>
</div>
<p class="caption">
  <span aria-hidden="true">(20)</span>
  <span class="visually-hidden">20 total reviews</span>
</p>
```

I was thinking about the icon contrast and high contrast theme issues; could we use CSS to display a half-star for a `.5` value instead of subdued? This would take care of both issues at the same time.

wiktoriaswiecicka commented:
> The subdued icon color does not meet WCAG 1.4.11: Non-text Contrast requirements of 3:1 (currently 1.35:1)

Is it critical that the disabled stars don't meet the contrast check? I feel like the enabled stars have to be much more visible and accessible. 

@svinkle I tried hollow outline and the component was more prominent and harder to scan for half-stars like below in the product grid. The opacity of text in contrary felt more like a light background for filled ones. This element is just there to enhance the design for those who can see it. Increasing contrast on the disabled stars can downgrade the experience for sighted users. 

<img width="464" alt="Screen Shot 2021-08-17 at 10 56 10 AM" src="https://user-images.githubusercontent.com/61301164/129749696-5138ac55-fa63-4208-ad3e-2762d1c0c64c.png">

svinkle commented:
@sofiamatulis There's also the the [`forced-colors` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) we could use to apply CSS specifically for the High Contrast theme without disrupting the original design.

sofiamatulis commented:
> @sofiamatulis There's also the the [`forced-colors` media query](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors) we could use to apply CSS specifically for the High Contrast theme without disrupting the original design.

@svinkle Amazing! I didn't know about this. What do you think of adding the stroke when this mode is on: https://github.com/Shopify/dawn/pull/341/commits/901ef69a308b8c5f6e2a2407325dca8b3c8754c0

Otherwise youd run into this: https://screenshot.click/17-42-ozqjy-ocl8j.png

svinkle commented:
@sofiamatulis I was trying out [system color values](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value#system_colors) (`CanvasText` with `GrayText` for example) but no matter what I try, it's always five white stars. I think it has to do with the `background` property not being well supported in High Contrast modes.

If you are able to add the stroke, let me know and I'll give it a shot.

sofiamatulis commented:
@gretchen-willenberg @melissaperreault Since we need to request translations by August 23rd, do you mind taking a look a the copy before we do so? :)

wiktoriaswiecicka commented:
A few small updates:

**Product card**
Font-size: 1.4rem
letter-spacing: 0.7

**Product page**
Vertical spacing: 1em change to 0.7rem
Font-size: 1.7rem
letter-spacing: 0.8

melissaperreault commented on locales/en.default.schema.json:
```diff
@@ -1282,6 +1291,9 @@
           "label": "Show vendor"
         },
+        "show_rating": {
+          "label": "Show product rating"
```
```suggestion
          "label": "Show product rating"
          "info":"Add a product rating app to display ratings for your products. [Learn more](https://help.shopify.com/en/manual/products/product-reviews/installation)"
```

melissaperreault commented on locales/en.default.schema.json:
```diff
@@ -1476,6 +1488,9 @@
           "label": "Show vendor"
+        },
+        "show_rating": {
+          "label": "Show product rating"
```
```suggestion
          "label": "Show product rating",
          "info": "Add a product rating app to display ratings for your products. [Learn more](https://help.shopify.com/en/manual/products/product-reviews/installation)"
```

gretchen-willenberg commented on locales/en.default.schema.json:
```diff
@@ -491,6 +491,9 @@
           "label": "Show vendor"
+        },
+        "show_rating": {
+          "label": "Show product rating"
```
Thank you, @melissaperreault ! Looking at that version along with the other settings, I can refine it to be singular and more like metafield support copy:
`To display a product rating, add a product ratings app. Learn more [offsite icon]`

If we say `product rating` too much since this is under `Show product rating`
`To display a rating, add a product ratings app. Learn more [offsite icon]`

In Sofia's [prototype video](https://screenshot.click/16-41-o5hev-b3q21.mp4), the block is called `Rating`; would adjust that to `Product rating`

melissaperreault replied:
Thank you Gretchen! 
I am aligned with this proposition, looks good to me! `To display a rating, add a product ratings app. Learn more [offsite icon]`

I made the edit in the suggestions too!

melissaperreault replied:
@gretchen-willenberg I was wondering if it should align with the metafield names, if not, aligned with matching merchants vocabulary that they are used to! Finally, the docs team can adapt with our preference!

melissaperreault commented on locales/en.default.schema.json:
```diff
@@ -1243,6 +1249,9 @@
           }
+        },
+        "rating": {
+          "name": "Rating"
```
```suggestion
          "name": "Product rating"
```
As per Gretchen's recommendation 🙏

melissaperreault commented on sections/main-product.liquid:
```diff
@@ -711,6 +732,11 @@
+    },
+    {
+      "type": "rating",
+      "name": "t:sections.main-product.blocks.rating.name",
```
@sofiamatulis We'll need to add the info text under the Product rating block inside the product template. There might be no setting but this information would benefit merchants here too!

melissaperreault replied:
Oh, in this case I believe `paragraph` would do the magic, no?

sofiamatulis replied:
@melissaperreault Thank you! Updated 🙏

sofiamatulis commented on sections/featured-collection.liquid:
```diff
@@ -152,6 +153,12 @@
+      "type": "checkbox",
+      "id": "show_rating",
+      "default": false,
+      "label": "t:sections.featured-collection.settings.show_rating.label"
```
@melissaperreault Since this is per section, I am adding the translations to all the sections :)

Here is the commit so you can see: https://github.com/Shopify/dawn/pull/341/commits/b55d21fe7cccc45151835abe6d72334ffc437375

sofiamatulis commented on locales/ko.schema.json:
```diff
@@ -1191,7 +1191,7 @@
             "show_dynamic_checkout": {
               "label": "동적 결제 버튼 표시",
-              "info": "스토어에서 사용 가능한 결제 방법을 사용하면 고객이 PayPal 또는 ApplePay처럼 원하는 옵션을 볼 수 있습니다. [자세히 알아보기](https://help.shopify.com/manual/using-themes/change-the-layout/dynamic-checkout)"
+              "info": "스토어에서 사용 가능한 결제 방법을 사용하면 고객이 PayPal 또는 ApplePay처럼 원하는 옵션을 볼 수 있습니다. [자세히 알아보기](https:\/\/help.shopify.com\/manual\/using-themes\/change-the-layout\/dynamic-checkout)"
```
@martinamarien I remember we ran into this issue on Kai's PR. Do you know why this is being added?

gretchen-willenberg commented on locales/en.default.schema.json:
```diff
@@ -961,6 +965,10 @@
         },
+        "show_rating": {
+          "label": "Show product rating",
+          "info": "To display a rating, add a product ratings app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"
```
Fact checking in the app store around "product ratings app" vs "product reviews app": 
- "product rating" returns 1465 app results
- "product reviews" returns 1523 results

Should we revise `product ratings app` to `product reviews app` in the text to capture more options?

"info": "To display a rating, add a product reviews app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"

gretchen-willenberg commented on locales/en.default.schema.json:
```diff
@@ -1324,6 +1340,10 @@
         },
+        "show_rating": {
+          "label": "Show product rating",
+          "info": "To display a rating, add a product ratings app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"
```
Fact checking in the app store around "product ratings app" vs "product reviews app": 
- "product rating" returns 1465 app results
- "product reviews" returns 1523 results

Should we revise `product ratings app` to `product reviews app` in the text to capture more options?

"info": "To display a rating, add a product reviews app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"

gretchen-willenberg commented on locales/en.default.schema.json:
```diff
@@ -1521,6 +1541,10 @@
+        },
+        "show_rating": {
+          "label": "Show product rating",
+          "info": "To display a rating, add a product ratings app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"
```
Fact checking in the app store around "product ratings app" vs "product reviews app": 
- "product rating" returns 1465 app results
- "product reviews" returns 1523 results

Should we revise `product ratings app` to `product reviews app` in the text to capture more options?

"info": "To display a rating, add a product reviews app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"

gretchen-willenberg reviewed (commented):
Fact checking in the app store around "product ratings app" vs "product reviews app": 
- "product rating" returns 1465 app results
- "product reviews" returns 1523 results

Should we revise `product ratings app` to `product reviews app` in the text to capture more options?

"info": "To display a rating, add a product reviews app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"

sofiamatulis commented:
> Everything looks great in terms of spacing, there is just one small tweak on the product card- margin-top: .0.7rem change to - 0.4rem.


@wiktoriaswiecicka Since this is being applied to all of them, do you want me to change it just for the rating or for all of them?

https://screenshot.click/19-23-b8q4t-d8qo6.mp4

wiktoriaswiecicka commented:
> > Everything looks great in terms of spacing, there is just one small tweak on the product card- margin-top: .0.7rem change to - 0.4rem.
> 
> @wiktoriaswiecicka Since this is being applied to all of them, do you want me to change it just for the rating or for all of them?
> 
> https://screenshot.click/19-23-b8q4t-d8qo6.mp4

Just for the rating block.

gretchen-willenberg commented on locales/en.default.schema.json:
```diff
@@ -509,6 +509,10 @@
+        },
+        "show_rating": {
+          "label": "Show product rating",
+          "info": "To display a rating, add a product ratings app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"
```
Would apply to each use: remove "s" from product rating**s** app

Final-final-final:
`To display a rating, add a product rating app. Learn more [offsite icon]`

gretchen-willenberg reviewed (changes requested):
Would apply to each use: remove "s" from product rating**s** app

Final-final-final:
`To display a rating, add a product rating app. Learn more [offsite icon]`

gretchen-willenberg reviewed (dismissed):
Thank you, Sofia!

sofiamatulis commented on locales/en.default.schema.json:
```diff
@@ -509,6 +509,10 @@
+        },
+        "show_rating": {
+          "label": "Show product rating",
+          "info": "To display a rating, add a product rating app. [Learn more](https:\/\/help.shopify.com\/en\/manual\/products\/product-reviews\/installation)"
```
@melissaperreault Thanks for the heads up 🙏 If we get the updated link before merge, I will replace it, but if not I can create a follow up issue

melissaperreault reviewed (dismissed):
I can approve this since it was already from UX, just see my note about the URL that would need to change once we have the right link!

KaichenWang commented on assets/component-rating.css:
```diff
@@ -0,0 +1,78 @@
+}
+
+.rating-star.color-icon-text {
+  --color-icon: rgb(var(--color-base-text));
```
Since the `.color-icon-XX` is determined by a theme setting, it would make sense to not limit the scope to only `.rating-star` (tho i am not sure where else we are planning to use this setting). We could simplify and move to `base.css`

```css
.color-icon-text {
  --color-icon: rgb(var(--color-base-text));
}

.color-icon-accent-1 {
  --color-icon: rgb(var(--color-base-accent-1));
}

.color-icon-accent-2 {
  --color-icon: rgb(var(--color-base-accent-2));
}

.color-icon-outline-button {
  --color-icon: rgb(var(--color-base-outline-button-labels));
}
```

sofiamatulis replied:
The reason that I kept it under there is because it isnt being used anywhere else 🤔 But this can def be reused in the future. Moved it 👍

KaichenWang reviewed (commented):
Looks good! Left some comments. Also, if we can revert the Shopify platform changes ([this one in particular](https://github.com/Shopify/dawn/pull/341/commits/73b300833f15e2dc8ea7a327eb21ceb7eec64651)), I will take another look at the final diff afterwards.

sofiamatulis commented on assets/component-card.css:
```diff
@@ -180,7 +180,13 @@
+  margin-top: 0.4rem;
+}
+
+.card-information__wrapper
```
@KaichenWang The ratings is only available on the product card (featured collection, prod recommendation, search results etc) and on the product page. Wik requested that we add some margin only on the product card and not on the product page.

sofiamatulis replied:
Right! So originally this is what we had on main:

```.card-information__wrapper > *:not(.visually-hidden:first-child) + * {
  margin-top: 0.7rem;
}
```

But I dont want this margin top when the child is a rating. So I added the `not`

sofiamatulis commented on sections/main-product.liquid:
```diff
@@ -307,6 +308,26 @@
+               assign rating_decimal = 0 
+               assign decimal = product.metafields.reviews.rating.value.rating | modulo: 1 
+               if decimal >= 0.3 and decimal <= 0.7
+                assign rating_decimal = 0.5
```
Good catch Kai 👍

KaichenWang reviewed (dismissed):
Looks good!

gretchen-willenberg reviewed (approved):
No changes to all uses of `To display a rating, add a product ratings app. Learn more`

sofiamatulis commented:
> No changes to all uses of `To display a rating, add a product ratings app. Learn more`

@gretchen-willenberg I have removed all `s`'s as per this comment: https://github.com/Shopify/dawn/pull/341#pullrequestreview-734471690

Is that correct? Here is the commit: https://github.com/Shopify/dawn/pull/341/commits/5effa758596fde3b2933ca7c53f12cdfd248b91c

tauthomas01 reviewed (dismissed):
LGTM 👍

melissaperreault commented on locales/en.default.json:
```diff
@@ -73,7 +73,10 @@
-    "loading": "Loading..."
+    "loading": "Loading...",
+    "total_reviews": "total reviews",
+    "star_reviews_info": "{{ rating_value }} out of {{ rating_max }} stars"
```
For context, are these buyer facing content on screen or via voice over?

sofiamatulis replied:
This is for the `aria-label`

KaichenWang reviewed (approved):
Looks good!

gretchen-willenberg commented:
> > No changes to all uses of `To display a rating, add a product ratings app. Learn more`
> 
> @gretchen-willenberg I have removed all `s`'s as per this comment: [#341 (review)](https://github.com/Shopify/dawn/pull/341#pullrequestreview-734471690)
> 
> Is that correct? Here is the commit: [5effa75](https://github.com/Shopify/dawn/commit/5effa758596fde3b2933ca7c53f12cdfd248b91c)

I'm sorry for copy-paste! Yes, removing the `Ss` was correct. 
`To display a rating, add a product rating app. Learn more`

sofiamatulis commented:
> Team, just noticing that it is currently not, but the `Star rating block` should also be included for the Featured product section, as a block, not installed by default.

@melissaperreault I have added it to the featured product 🙏 good catch

notrealdev commented:
Hi, how to get reviews count?

For eg. If I wanted to display the following somewhere on my website:

Code:
`This is an awesome product. So far {{product.reviews_count}} people have reviewed this product.`

Output:
`This is an awesome product. So far 54 people have reviewed this product.`
